einsum.lisp

```diff
@@ -0,0 +1,194 @@
+    (mapcar #'calc-dim output-indices)))
+
+
+;;; To evaluate the einsum expression, the %EINSUM function below
+;;; generates a bunch of nested DOTIMES loops.  This could be improved
+;;; in at least three possible directions:
+;;; 
+;;; 1. This top-level EINSUM macro could look at the size of the
+;;;    involved arrays, and generate several smaller calls to %EINSUM
+;;;    which are subsequently combined, akin to the dynamic programming
+;;;    approach to matrix multiplication.
+;;; 2. Within a single instance of %einsum, there is room for improving
+;;;    the ordering of indices in our loops, or perhaps even blocking for
+;;;    better cache usage.
+;;; 3. Parts of this could be parallelized.
+
+
+(defun %einsum (output-array output-indices &rest factors)
+  "Translates an einsum expression into corresponding iterative code
+to perform the summation, updating and returning OUTPUT-ARRAY. If
+OUTPUT-ARRAY is null, a newly allocated array is used.
+
+The einsum expression contains a number of FACTORS, which are
+expressions of the form (A i1 i2 ... il) where A is a k-dimensional
+array and i1, i2, ..., ik are symbols denoting indices. For the
+expression to be meaningful, each index should appear exactly twice,
+either paired with another index in FACTORS, or paired with an index
+in OUTPUT-INDICES, and the associated arrays should have the same size
+along paired indices. "
+
+  ;; 
+  (let* ((index-table (apply #'build-index-table factors))
+         (output-dims (loop :for i :from 0 :below (length output-indices) :collecting (gensym "dim")))
+         (summation-indices (loop :for idx :being :the :hash-keys :of index-table
+                                  :for entries :being :the :hash-values :of index-table
+                                  :when (= 2 (length entries))
+                                    :collect idx))
+         (summation-dims (loop :for i :from 0 :below (length summation-indices) :collecting (gensym "dim")))
+         (result-array (gensym "result")) ; array to store results
+         (sum-var (gensym)))            ; updated in the innermost nested loop
+
+    ;; The result of all of this is basically a few nested loops
+    ;; enclosed in some let expressions which introduce local
+    ;; variables. We build this "from the inside out", by starting
+    ;; with what goes in the innermost loop (INNER-SUM-TALLY) below,
+    ;; and then defining functions to successsively wrap this
+    ;; expression accordingly.
+    (labels ((inner-sum-tally ()
+               "Compute the product of factors and update the summand."
+               `(incf ,sum-var (* ,@ (mapcar #'factor-aref-expr factors))))
+             (indices-loop (indices dims expr)
+               "Wrap EXPR in nested loops, one for each element of INDICES."
+               (loop :for idx :in indices
+                     :for dim :in dims
+                     :do (setf expr `(dotimes (,idx ,dim)
+                                         ,expr)))
```

Comment:
move this onto the prev line

Reply:
should you declare `idx` ignorable? is it possible `idx` won't be used in the expression?

```
(declare (ignorable ,idx))
```

this comment applies throughout the file for generated `let`s and `doX`s

Reply:
In this instance, unused indices would be a bug, as every index is either i) something being summed over, in which case it is used to refer to specific elements of the input arrays, or ii) an index to the output array, in which case it is used when we fill in the summed values from (i).